api/src/routes/template-v2.ts

```diff
@@ -0,0 +1,185 @@
+        name,
+        templateId,
+        status: 204,
+      });
+
+      res.status(204);
+      return res.send(
+        `GET /templates/:template_id -- Did not find template for template_id=${templateId}`
+      );
+    }
+
+    mixpanelTrack("get_template_by_name", {
+      name,
+      templateId,
+      status: 200,
+    });
+
+    return res.send(template);
+  });
+
+  router.get("/templates/manifest", async (req: Request, res: Response) => {
+    const templateManifest = await templateService.getTemplateManifest();
+
+    if (!templateManifest) {
+      mixpanelTrack("get_template_manifest", {
+        status: 204,
+      });
+
+      res.status(204);
+      return res.send(
+        `GET /templates/manifest -- Did not find template manifest`
+      );
+    }
+
+    mixpanelTrack("get_template_manifest", {
+      status: 200,
+    });
+
+    return res.send(templateManifest);
+  });
+
+  router.get("/templates/:template_id", async (req: Request, res: Response) => {
+    const templateId = req.params.template_id;
+
+    const template = await templateService.getTemplate(templateId);
+
+    if (!template) {
+      mixpanelTrack("get_template", {
+        templateId,
+        status: 204,
+      });
+
+      res.status(204);
+      return res.send(
+        `GET /templates/:template_id -- Did not find template for template_id=${templateId}`
+      );
+    }
+
+    mixpanelTrack("get_template", {
+      templateId,
+      status: 200,
+    });
+
+    return res.send(template);
+  });
+
+  router.post("/templates/search/messages", async (req: Request, res: Response) => {
+    const page = req.body.page as number || undefined;
+    const range = req.body.range as number || undefined;
+    const searchMessagesTitleENUS = req.body.searchMessagesTitleENUS as string || undefined;
+    const searchMessagesDescriptionENUS = req.body.searchMessagesDescriptionENUS as string || undefined;
+
+    try {
+      const templates = await templateService.searchTemplates(
+        page ?? 0,
+        range ?? 100,
+        searchMessagesTitleENUS,
+        searchMessagesDescriptionENUS
+      )
+      return res.send(templates)
+
+    } catch (e) {
+      mixpanelTrack("search_template_messages", {
+        page,
+        range,
+        searchMessagesTitleENUS,
+        searchMessagesDescriptionENUS,
+        status: 400,
+      });
+      res.status(400);
+      return res.send("POST /templates/search/messages -- Error occurred when getting template");
+    }
+  })
+
+  router.post("/templates/search/cadence", async (req: Request, res: Response) => {
+    const cadence_base64 = req.body.cadence_base64 as string;
+    const network = req.body.network as string
+
+    let cadence = Buffer.from(cadence_base64, "base64").toString("utf8");
+    let cadenceAST = await parseCadence(cadence);
+
+    let template;
+    try {
+      template = await templateService.getTemplateByCadenceASTHash(
+        await genHash(cadenceAST),
```

Comment:
You're awaiting the `genHash` function multiple times for the same input. Compute it once.